src/nixenv.rs

```diff
@@ -0,0 +1,545 @@
+
+            let intersection = candidates
+                .intersection(&all_new_paths)
+                .cloned()
+                .collect::<Vec<PathBuf>>();
+            if !intersection.is_empty() {
+                return Err(NixEnvError::PathConflict(pkg.to_path_buf(), intersection));
+            }
+
+            all_new_paths.extend(candidates.into_iter());
+        }
+
+        let tmp = tempfile::tempdir().map_err(NixEnvError::CreateTempDir)?;
+        let temporary_profile = tmp.path().join("profile");
+
+        // Construct an empty profile
+        {
+            let output = tokio::process::Command::new(self.nix_store_path.join("bin/nix"))
+                .process_group(0)
+                .set_nix_options(self.nss_ca_cert_path)?
+                .args([
+                    "build",
+                    "--expr",
+                    r#"
+                    derivation {
+                        name = "user-environment";
+                        system = "builtin";
+                        builder = "builtin:buildenv";
+                        derivations = [];
+                        manifest = builtins.toFile "env-manifest.nix" "[]";
+                    }
+                "#,
+                    "--out-link",
+                ])
+                .arg(&temporary_profile)
+                .output()
+                .await
+                .map_err(|e| NixEnvError::StartNixCommand("nix build-ing an empty profile", e))?;
+
+            if !output.status.success() {
+                return Err(NixEnvError::NixCommand(
+                    "nix build-ing an empty profile",
+                    output,
+                ));
+            }
+        }
+
+        if let Ok(canon_profile) = self.profile.canonicalize() {
+            tracing::info!("Duplicating the existing profile into the scratch profile");
+
+            let output = tokio::process::Command::new(self.nix_store_path.join("bin/nix-env"))
+                .process_group(0)
+                .arg("--profile")
+                .arg(&temporary_profile)
+                .arg("--set")
+                .arg(canon_profile)
+                .output()
+                .await
+                .map_err(|e| NixEnvError::StartNixCommand("nix build-ing an empty profile", e))?;
+
+            if !output.status.success() {
+                return Err(NixEnvError::NixCommand(
+                    "nix build-ing an empty profile",
+                    output,
+                ));
+            }
+        }
+
+        // Query packages that are already installed in the profile.
+        // Constructs a map of (store path in the profile) -> (hash set of paths that are inside that store path)
+        let mut installed_paths: HashMap<PathBuf, HashSet<PathBuf>> = HashMap::new();
+        {
+            let output = tokio::process::Command::new(self.nix_store_path.join("bin/nix-env"))
+                .process_group(0)
+                .arg("--profile")
```

Comment:
This may be problematic for any time we operate on profiles that may have been touched by the `nix profile` command:

```
$ nix-env --query --installed --out-path --json
error: profile '/home/vin/.local/state/nix/profiles/profile' is incompatible with 'nix-env'; please use 'nix profile' instead
```

(but we already have this problem, so maybe-fine to not deal with it now)